### PR TITLE
[libc][math][c23] Add exp10f16 C23 math function

### PR DESCRIPTION
**Pull request by overmighty**
Part of #95250.

### CODE REVIEW

**Comment by overmighty** [comment]
Based on #101217.

File: libc/src/math/generic/exp2f16.cpp [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    {0xaf57U, 0x3b63U, 1U, 0U, 0U},
+}};
+
+// Generated by Sollya with the following commands:
+//   > display = hexadecimal;
+//   > for i from 0 to 7 do printsingle(round(2^(i * 2^-3), SG, RN));
+static constexpr cpp::array<uint32_t, 8> EXP2_MID_BITS = {
+    0x3f80'0000U, 0x3f8b'95c2U, 0x3f98'37f0U, 0x3fa5'fed7U,
+    0x3fb5'04f3U, 0x3fc5'672aU, 0x3fd7'44fdU, 0x3fea'c0c7U,
+};
```
Comment (overmighty) [inline libc/src/math/generic/exp2f16.cpp:42]: Copy-pasted from exp2f16.cpp. I guess I should create a shared expxf16.h header instead.

File: libc/src/math/generic/exp10f16.cpp [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+//   > round(log10(2), SG, RN);
+static constexpr float LOG10F_2 = 0x1.344136p-2f;
+
+// Generated by Sollya with the following commands:
+//   > display = hexadecimal;
+//   > for i from 0 to 7 do printsingle(round(2^(i * 2^-3), SG, RN));
+static constexpr cpp::array<uint32_t, 8> EXP2_MID_BITS = {
+    0x3f80'0000U, 0x3f8b'95c2U, 0x3f98'37f0U, 0x3fa5'fed7U,
+    0x3fb5'04f3U, 0x3fc5'672aU, 0x3fd7'44fdU, 0x3fea'c0c7U,
+};
```
Comment (overmighty) [inline libc/src/math/generic/exp10f16.cpp:71]: Copy-pasted from exp2f16.cpp. I guess I should create a shared expxf16.h header instead.

**Comment by overmighty** [comment]
On i7-13700H:

- With Clang 18, without `-march=native`:  
  ```
   Performance tests with inputs in denormal range:
  -- My function --
       Total time      : 1627880300 ns 
       Average runtime : 79.564 ns/op 
       Ops per second  : 12568491 op/s 
  -- Other function --
       Total time      : 21792295 ns 
       Average runtime : 1.06512 ns/op 
       Ops per second  : 938863942 op/s 
  -- Average runtime ratio --
       Mine / Other's  : 74.6998 
  
   Performance tests with inputs in normal range:
  -- My function --
       Total time      : 25289960974 ns 
       Average runtime : 41.1634 ns/op 
       Ops per second  : 24293434 op/s 
  -- Other function --
       Total time      : 648779538 ns 
       Average runtime : 1.05599 ns/op 
       Ops per second  : 946978078 op/s 
  -- Average runtime ratio --
       Mine / Other's  : 38.9808 
  ```

- With Clang 18, with `-march=native`:  
  ```
   Performance tests with inputs in denormal range:
  -- My function --
       Total time      : 348120098 ns 
       Average runtime : 17.0147 ns/op 
       Ops per second  : 58772820 op/s 
  -- Other function --
       Total time      : 21815025 ns 
       Average runtime : 1.06623 ns/op 
       Ops per second  : 937885700 op/s 
  -- Average runtime ratio --
       Mine / Other's  : 15.9578 
  
   Performance tests with inputs in normal range:
  -- My function --
       Total time      : 24294542949 ns 
       Average runtime : 39.5432 ns/op 
       Ops per second  : 25288806 op/s 
  -- Other function --
       Total time      : 648768173 ns 
       Average runtime : 1.05597 ns/op 
       Ops per second  : 946994667 op/s 
  -- Average runtime ratio --
       Mine / Other's  : 37.4472 
  ```

File: libc/src/math/generic/exp10f16.cpp [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+#include "src/__support/FPUtil/rounding_mode.h"
+#include "src/__support/common.h"
+#include "src/__support/macros/config.h"
+#include "src/__support/macros/optimization.h"
+#include "src/__support/macros/properties/cpu_features.h"
+
+namespace LIBC_NAMESPACE_DECL {
+
+static constexpr size_t N_EXP10F16_EXCEPTS = 5
+#ifndef LIBC_TARGET_CPU_HAS_FMA
```
Comment (lntue) [inline libc/src/math/generic/exp10f16.cpp:29]: This is actually harder to read and looks a bit odd.  Just do the normal ifdef switch:
```
#ifdef LIBC_TARGET_CPU_HAS_FMA
static constexpr size_t N_EXP10F16_EXCEPTS = 5;
#else
static constexpr size_t N_EXP10F16_EXCEPTS = 8;
#endif
```

**Review by lntue — APPROVED** [review]
LGTM with a nit.

**Pull request merged**